File: Programmers/Lv.1/숫자 짝꿍/숫자 짝꿍.py
def solution(X, Y):
    nums = {}
    common = []
    
    # X를 구성하는 숫자 dict로 정리
    for n in list(X):
        if n in nums:
            nums[n] += 1
        else:
            nums[n] = 1
            
    # Y를 순회하면서 공통 여부 체크
    for n in list(Y):
        if n in nums and nums[n] > 0:
            common.append(n)
            nums[n] -= 1          
            
    if common:
        if common[0] == '0':
            return '0'
        return ''.join(sorted(common, reverse=True))
    else:
        return '-1'

File: Programmers/Lv.1/숫자 짝꿍/숫자 짝꿍.java
import java.util.*;

class Solution {
    public String solution(String X, String Y) {
        String answer = "";
        
        HashMap<Character, Integer> nums = new HashMap<>();
        
        // X에 있는 숫자들의 갯수를 해시맵에 저장
        for (int i=0; i < X.length(); i++) {
            char curr = X.charAt(i);
            int value = 0;
            
            if (nums.containsKey(curr)) {
                value = nums.get(curr);
            }
            nums.put(curr, value+1);
        }
        
        List<Character> common = new ArrayList<>();
        
        // Y에 있는 숫자들을 보면서 겹치는 숫자를 기록
        for (int i=0; i<Y.length(); i++) {
            char curr = Y.charAt(i);
            int value;
            
            if (nums.containsKey(curr)) {
                value = nums.get(curr);
                if (value > 0) {
                    common.add(curr);
                }
                nums.put(curr, value - 1);
            }
            
        }
        
        if (common.size() == 0)
            return "-1";
        
        common.sort(Comparator.reverseOrder());
        if (common.get(0) == '0')
            return "0";
    
        StringBuilder sb = new StringBuilder();
        for (Character ch: common) {
            sb.append(ch);
        }
        
        return sb.toString();
    }
}